test/test_sparse.py

```diff
@@ -3272,6 +3272,20 @@ def test_sparse_consistency(self, device, dtype, op):
         assert torch.is_tensor(output)
         self.assertEqual(output.to_dense(), expected)
 
+        inplace_variant = op.get_inplace()
+
+        if inplace_variant is not None:
+            sparse_tensor_copy = sample.to_sparse().clone()
+            if torch.can_cast(output.dtype, sparse_tensor_copy.dtype) is False:
```

Comment:
Use `if not x:` instead of `if x is False`. If `x` can be a non-boolean object (say, `None`), then use explicit checks.

Reply:
Or better yet, use:
```
if torch.can_cast(output.dtype, sparse_tensor_copy.dtype) and (
      sparse_tensor_copy.is_coalesced() or op.is_linear_map):
    inplace_variant(sparse_tensor_copy)
    self.assertEqual(expected, sparse_tensor_copy.to_dense())
else:
    with self.assertRaises(RuntimeError):
￼        inplace_variant(sparse_tensor_copy)
```

Reply:
Hmm, since `sparse_tensor_copy = sample.to_sparse().clone()`, isn't `sparse_tensor_copy` always coalesced?
If so, the `(.. or ..)` expression is always True and can be eliminated.

In fact, the logic in `sparse_unary_op_check_and_init` would automatically coalesce uncoalesce inputs for non-linear maps and there would be no RuntimeError. This indicates that the current test-suite does not cover uncoalesced inputs. I suggest defining:
```
class TestUncoalescedSparseUnaryUfuncs(TestSparseUnaryUfuncs):
    def setUp(self):
        super(TestUncoalescedSparseUnaryUfuncs, self).setUp()
        self.is_uncoalesced = True
```

Reply:
> Or better yet, use:
> 
> ```
> if torch.can_cast(output.dtype, sparse_tensor_copy.dtype) and (
>       sparse_tensor_copy.is_coalesced() or op.is_linear_map):
>     inplace_variant(sparse_tensor_copy)
>     self.assertEqual(expected, sparse_tensor_copy.to_dense())
> else:
>     with self.assertRaises(RuntimeError):
> ￼        inplace_variant(sparse_tensor_copy)
> ```

@pearu - Have done the changes as suggested

Reply:
> Hmm, since `sparse_tensor_copy = sample.to_sparse().clone()`, isn't `sparse_tensor_copy` always coalesced?
> If so, the `(.. or ..)` expression is always True and can be eliminated.
> 
> In fact, the logic in `sparse_unary_op_check_and_init` would automatically coalesce uncoalesce inputs for non-linear maps and there would be no RuntimeError. This indicates that the current test-suite does not cover uncoalesced inputs. I suggest defining:
> 
> ```
> class TestUncoalescedSparseUnaryUfuncs(TestSparseUnaryUfuncs):
>     def setUp(self):
>         super(TestUncoalescedSparseUnaryUfuncs, self).setUp()
>         self.is_uncoalesced = True
> ```

@pearu - I agree that currently, this script is not testing for uncoalesced tensors; the reason being we are generating sample inputs as strided tensors. For testing uncoalesced tensors; we need to create sample inputs as sparse tensors; which a WIP at my end and is going to take some time.
If you suggest I can remove the (.. or ..) for the time being to avoid the confusion; and raise it along with uncoalesced testing as a separate PR

Reply:
@hlahkar it is fine to postpone the uncoalesced tensors testing to a follow-up PR but since we know that the logic could be wrong here, let us fix it now. For that, try running the tests *locally* by inserting the following statements before the `if torch.can_cast..` block:
```
sparse_tensor_copy = sparse_tensor_copy[0]
expected = expected[0]
```
so that the `sparse_tensor_copy` will be uncoalesced (use `assert not sparse_tensor_copy.is_coaleasced()` to be sure).

I suspect that `(... or ...)` can be removed in general but before removing it, check it locally.